### PR TITLE
Feature: Support for UserConsentPage

### PR DESCRIPTION
**Pull request by xzxiaoshan**
1. The current authorization consent page is fixed by splicing HTML in the code (unable to extend the style)

2. This modification abstracts the UserConsentPage as an interface, which can be customized to achieve the purpose of customized page

3. On the basis of modifying a small amount of code, users can implant their own customized pages (for example, it can be generated by freemarker/thymeleaf, etc.)

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@xzxiaoshan Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects-experimental/spring-authorization-server&pullRequestId=270)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects-experimental/spring-authorization-server&pullRequestId=270) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by lanmingle** [comment]
recommendation:

remove UserConsentPage interface default generate method , enable util class or impl class 。

eg:

UserConsentPage -> GenerateUserConsentPage
UserConsentPage -> RedirectUserConsentPage (enable queryParam, eg: redirect nginx example.com/us?p1=&p2 ...)

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+	default boolean isConsentApproved(HttpServletRequest request) {
+		return CONSENT_ACTION_APPROVE.equalsIgnoreCase(request.getParameter(CONSENT_ACTION_PARAMETER_NAME));
+	}
+
+	default boolean isConsentCancelled(HttpServletRequest request) {
+		return CONSENT_ACTION_CANCEL.equalsIgnoreCase(request.getParameter(CONSENT_ACTION_PARAMETER_NAME));
+	}
+
+	default String generateConsentPage(HttpServletRequest request,
```
Comment (lanmingle) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java:47]: enable util class or impl class ?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+ * @date 2021/3/30 15:42
+ */
+public interface UserConsentPage {
+
+	MediaType TEXT_HTML_UTF8 = new MediaType("text", "html", StandardCharsets.UTF_8);
+	String CONSENT_ACTION_PARAMETER_NAME = "consent_action";
+	String CONSENT_ACTION_APPROVE = "approve";
+	String CONSENT_ACTION_CANCEL = "cancel";
+
+	default void displayConsent(HttpServletRequest request, HttpServletResponse response,
```
Comment (lanmingle) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java:30]: remove UserConsentPage interface default method , enable util class or impl class 。

Comment (xzxiaoshan) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java:30]: OK，Implementation of DefaultGenerateUserConsentPage

**Review by xzxiaoshan — COMMENTED** [review]
From Guangdong (Location prompt) ? Are you familiar with Chinese?
1、本来打算写一个默认的impl（DefaultGenerateUserConsentPage），这个我还是按照impl这样修改一下。
2、这个地方不太推荐使用自定义URL重定向的方式来实现，因为这个页面的呈现或提交逻辑很简单比较清晰、正常来说只有样式变更需求不需要进行更多的数据扩展，如果采用重定向这样的方式，对扩展来说还需要做好“同意”提交的处理会增加麻烦（现在只只修改内容，也比较符合这个UserConsentPage的概念，对提交的处理逻辑也不存在问题都已经交由底层统一处理了）

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java (hunk 2 above)
Comment (xzxiaoshan) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java]: 1. isConsentApproved() and isConsentCancelled() retain default.
2. displayConsent() and generateConsentPage() impl class.

**Comment by pivotal-issuemaster** [comment]
@xzxiaoshan Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects-experimental/spring-authorization-server&pullRequestId=270)!

**Comment by lanmingle** [comment]
@xzxiaoshan 保持 UserConsentPage 接口，实现一个 DefaultGenerateUserConsentPage 默认类，是否提供重定向这个后续使用可以自己扩展实现，一些做产品的，别人使用一些自定义页面这个很常有，尽量做到前后端分离。

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+	/**
+	 * Generate Consent Page HTML Code
+	 *
+	 * @param request
+	 * @param registeredClient
+	 * @param authorization
+	 * @return HTML Code
+	 */
+	String generateConsentPage(HttpServletRequest request,
```
Comment (lanmingle) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java:78]: 不建议这个方法，这个方法应该由实现自行处理，入口是基于 displayConsent 方法。

DefaultGenerateUserConsentPage：

``` java
        
        // 注意 @Override 添加注解
        @Override
	public void displayConsent(HttpServletRequest request, HttpServletResponse response,
			RegisteredClient registeredClient, OAuth2Authorization authorization) throws IOException {
                // 对于写入 html 代码，可选两种，1是：抽象一个 AbsGenerateUserConsentPage（偷懒）
                // 2是：搞个 StaticGenerateUserConsentPage 通过调用注入 handler 委托实现，如
                // String consentPage = this.handler.generateConsentPage(request, registeredClient, authorization);
		String consentPage = this.generateConsentPage(request, registeredClient, authorization);
		response.setContentType(TEXT_HTML_UTF8.toString());
		response.setContentLength(consentPage.getBytes(StandardCharsets.UTF_8).length);
		response.getWriter().write(consentPage);
	}

	protected String generateConsentPage(HttpServletRequest request,
			RegisteredClient registeredClient, OAuth2Authorization authorization) {

```

Comment (xzxiaoshan) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java:78]: 1、这个default是默认的，保留原有的处理方式。原作者之所以默认拼接HTML渲染了一个页面，是为了让程序可以跑起来正常使用。
2、对于真正的项目业务来说，这个DefaultGenerateUserConsentPage是没有人会使用的，一定会自定义的。而在自定义实现的时候，displayConsent 和 generateConsentPage 都是需要用户必须重写的（自然是用户自行处理）你对这个默认的UserConsentPage的理解可能有些偏差。
3、UserConsentPage 不是完全新增的东西，是从之前的Filter类里面的static内部类提取出来的，保留了默认的方法实现结构。

Comment (xzxiaoshan) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java:78]: 对于 UserConsentPage 的扩展实现，代码都有的，你可能没有看全。扩展的实现是在这个jar发布后之外的Configuration里面配置的。就是因为原来的 UserConsentPage 是内部类无法扩展自定义，才有了这个接口的抽象和feature。

Comment (xzxiaoshan) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java:78]: /**
 * Custom consent authorization return page
 *
 * @author shanhy
 * @date 2021/3/30 15:40
 */
@Configuration(proxyBeanMethods = false)
public class CustomOAuth2AuthorizationServerConfiguration {

	@Bean
	@ConditionalOnMissingBean
	@Order(Ordered.HIGHEST_PRECEDENCE)
	public SecurityFilterChain authorizationServerSecurityFilterChain(HttpSecurity http) throws Exception {
		applyDefaultSecurity(http);
		return http.build();
	}

	// @formatter:off
	public static void applyDefaultSecurity(HttpSecurity http) throws Exception {
		OAuth2AuthorizationServerConfigurer<HttpSecurity> authorizationServerConfigurer =
				new OAuth2AuthorizationServerConfigurer<>();

		// Setting CustomUserConsentPage（这里设置自定义的页面）
		authorizationServerConfigurer.setUserConsentPage(new CustomGenerateUserConsentPage());

               （后续代码，略）

Comment (lanmingle) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/UserConsentPage.java:78]: 好！稍后有空我把代码克隆下来细看一下。👍

**Comment by jgrandja** [comment]
Thanks for the PR @xzxiaoshan.

It was never the intention to expose `UserConsentPage` as a public API. It's strictly meant for internal implementation only.

The plan for allowing a custom consent page to be configured was to implement the same pattern as `HttpSecurity.formLogin().loginPage()`.

We have a PR in progress gh-280 that has implemented this way and additionally provided a mechanism for saving/loading user consent decisions, which is needed as part of this feature/customization.

As an FYI, it's always best to reach out first before you start any work so we can align on approach and to also ensure the task is not already taken.

I'm going to close this in favour of gh-280

**Pull request closed without merging**